scipy/spatial/transform/rotation.py

```diff
@@ -1639,3 +1640,308 @@ def __call__(self, times):
+    result[~non_zero] = [0, 0, 0]
+    return result
+
+
+def _spline_w_i_modifier(rotvec, omega):
+    # B function formulation from eqn 24
+    # rotvec and omega should have same shape
+    dtheta = np.linalg.norm(rotvec, axis=1)
+    non_zero = (dtheta != 0)
+
+    e = np.zeros(rotvec.shape)
+    e[non_zero] = rotvec[non_zero] / dtheta[non_zero][:, None]
+
+    dot_prod = np.einsum('ij,ij->i', omega, e)[:, None]
+    cross_prod = np.cross(e, omega)
+
+    small_theta = (dtheta <= 1e-6)
+    small_angle = dtheta[small_theta, None]
+    large_angle = dtheta[~small_theta, None]
+
+    c1 = np.empty((dtheta.shape[0], 1))
+    c1[~small_theta] = np.sin(large_angle) / large_angle
+    c1[small_theta] = 1 - (small_angle ** 2) / 6 + (small_angle ** 4) / 120
+
+    c2 = np.empty_like(c1)
+    c2[~small_theta] = (1 - np.cos(large_angle)) / large_angle
+    c2[small_theta] = small_angle / 2 - (small_angle ** 3) / 24
+
+    result = dot_prod * e + c1 * np.cross(cross_prod, e) - c2 * cross_prod
+    result[~non_zero] = omega[~non_zero]
+
+    return result
+
+
+def _spline_w_f_modifier(rotvec, omega):
+    # B inverse function formulation from eqn 26
+    # rotvec and omega should have same shape
+    dtheta = np.linalg.norm(rotvec, axis=1)
+    non_zero = (dtheta != 0)
+
+    e = np.zeros(rotvec.shape)
+    e[non_zero] = rotvec[non_zero] / dtheta[non_zero][:, None]
+
+    dot_prod = np.einsum('ij,ij->i', e, omega)[:, None]
+    cross_prod = np.cross(e, omega)
+
+    small_theta = (dtheta <= 1e-6)
+    small_angle = dtheta[small_theta, None]
+    large_angle = dtheta[~small_theta, None]
+
+    c1 = np.empty((dtheta.shape[0], 1))
+    c1[~small_theta] = large_angle / (2 * np.tan(large_angle))
+    c1[small_theta] = 1 - (small_angle ** 2) / 12 - (small_angle ** 4) / 720
+
+    c2 = np.empty_like(c1)
+    c2[~small_theta] = (1 - np.cos(large_angle)) / large_angle
+    c2[small_theta] = small_angle / 2 - (small_angle ** 3) / 24
+
+    return (dot_prod * e + c1 * np.cross(cross_prod, e) +
+            dtheta[:, None] * cross_prod / 2)
+
+
+def _intermediate_step(w_prev, rotvecs, dt_inv, w_i, w_f):
+    # Page 7 of reference. Convection on delta time has changes from here.
+    # only the first a value is required
+    a2 = 2 * dt_inv[0]
+    b = 4 * (dt_inv[:-1] + dt_inv[1:])
+    c = 2 * dt_inv[1:]
+
+    d = (6 * rotvecs[:-1] * (dt_inv[:-1, None] ** 2) +
+         6 * rotvecs[1:] * (dt_inv[1:, None] ** 2) -
+         _spline_non_linear_term(rotvecs[:-1], w_prev))
+    d[0] -= a2 * (_spline_w_i_modifier(rotvecs[0, None], w_i[None, :]))[0]
+    d[-1] -= c[-1] * (_spline_w_f_modifier(rotvecs[-1, None], w_f[None, :]))[0]
+
+    w = np.empty_like(w_prev)
+    w[-1] = d[-1] / b[-1]
+    w[:-1] = ((d[:-1] -
+              c[:-1, None] * _spline_w_f_modifier(rotvecs[1:-1], w_prev[1:])) /
+              b[:-1, None])
+
+    error = w - w_prev
+    diff = np.sum(np.dot(error, error.T))
+
+    return w, diff
+
+
+class QSpline(object):
```

Comment:
How about calling this `QuaternionSpline`? Or `QuaternionInterpolator`? Or yet something else - just want to avoid giving the impression that this is a spline thingy that can be reused (it's very similar to B-spline now).

Reply:
How about renaming both functions to `RotationSlerp` and `RotationSpline`? In my mind this makes is explicit that these classes are for use only with `Rotation`s, as opposed to general interpolators for quaternions.

Reply:
`RotationSlerp` and `RotationSpline` sounds logical to me as well.